indra/newview/llluamanager.cpp

```diff
@@ -401,3 +401,213 @@ void LLLUAmanager::runScriptOnLogin()
+
+    if (is_absolute_path(mPathToResolve))
+        return moduleStatus;
+
+    std::vector<std::string> lib_paths;
+
+    lib_paths.push_back(gDirUtilp->getExpandedFilename(LL_PATH_APP_SETTINGS, ""));
+
+    for (size_t i = 0; i < lib_paths.size(); ++i)
+    {
+        std::string absolutePathOpt = join_paths(lib_paths[i], mPathToResolve);
+
+        if (absolutePathOpt.empty())
+            luaL_errorL(L, "error requiring module");
+
+        mChunkname = absolutePathOpt;
+        mAbsolutePath = absolutePathOpt;
+
+        moduleStatus = findModuleImpl();
+
+        if (moduleStatus != LLRequireResolver::ModuleStatus::NotFound)
+            return moduleStatus;
+    }
+
+    return LLRequireResolver::ModuleStatus::NotFound;
+}
+
+LLRequireResolver::ModuleStatus LLRequireResolver::findModuleImpl()
+{
+    std::string possibleSuffixes[] = {".luau", ".lua"};
+
+    size_t unsuffixedAbsolutePathSize = mAbsolutePath.size();
+
+    for (auto possibleSuffix : possibleSuffixes)
+    {
+        mAbsolutePath += possibleSuffix;
+
+        // Check cache for module
+        lua_getfield(L, -1, mAbsolutePath.c_str());
+        if (!lua_isnil(L, -1))
+        {
+            return ModuleStatus::Cached;
+        }
+        lua_pop(L, 1);
+
+        // Try to read the matching file
+        std::string source = read_file(mAbsolutePath);
+        if (!source.empty())
+        {
+            mChunkname  = "=" + mChunkname + possibleSuffix;
+            mSourceCode = source;
+            return ModuleStatus::FileRead;
+        }
+
+        mAbsolutePath.resize(unsuffixedAbsolutePathSize);  // truncate to remove suffix
+    }
+
+    return ModuleStatus::NotFound;
+}
+
+void LLRequireResolver::resolveAndStoreDefaultPaths()
+{
+    if (!is_absolute_path(mPathToResolve))
+    {
+        std::string path = gDirUtilp->getDirName(mSourceChunkname);
+        std::replace(path.begin(), path.end(), '\\', '/');
+        path = join_paths(path, mPathToResolve);
+        mAbsolutePath = path;
+        mChunkname = path;
+    }
+    else
+    {
+        mChunkname = mPathToResolve;
+        mAbsolutePath = mPathToResolve;
+    }
+}
+
+static int finishrequire(lua_State *L)
+{
+    if (lua_isstring(L, -1))
+        lua_error(L);
+
+    return 1;
+}
+
+lua_function(require, "require(module_name) : module_name can be fullpath or just the name, in both cases without .lua")
+{
+    std::string name = luaL_checkstring(L, 1);
+
+    LLRequireResolver::ResolvedRequire resolvedRequire = LLRequireResolver::resolveRequire(L, std::move(name));
+
+    if (resolvedRequire.status == LLRequireResolver::ModuleStatus::Cached)
+        return finishrequire(L);
+    else if (resolvedRequire.status == LLRequireResolver::ModuleStatus::NotFound)
+        luaL_errorL(L, "error requiring module");
+
+    // module needs to run in a new thread, isolated from the rest
+    // note: we create ML on main thread so that it doesn't inherit environment of L
+    lua_State *GL = lua_mainthread(L);
```

Comment:
My `listen_events()` function wanted to find the main thread corresponding to the calling thread, which is why it originally referenced `LUA_RIDX_MAINTHREAD`. I didn't know about `lua_mainthread()`, and in fact I still can't find any documentation about it. Where do you find these? By reading the luau source?

It wouldn't have occurred to me to load a module on its own thread, but it makes sense not to let it accidentally reference globals defined by the caller of `require()`. But why is `lua_newthread(lua_mainthread(L))` better than `lua_newstate()`?

Reply:
Documentation is lacking. Current `require()` implementation is heavily based on [luau CLI implementation.](https://github.com/luau-lang/luau/blob/d6c2472f0c5c946a12ad54bb234747f62e15458f/CLI/Repl.cpp#L112-L245)